0000-classical-stores.md

```diff
@@ -0,0 +1,493 @@
+The `location` must evaluate to a writeable memory location.
+In this initial RFC this is limited to `Var` instances, though in the future one can imagine some `expr.Index` node or the like existing, for runtime indexing into objects that are array-like in bits.
+
+`QuantumCircuit.get_var` and `.has_var` are convenience methods for the user.
+`.has_var` is a simple Boolean checker akin to `dict.__contains__`.
+`.get_var` is akin to `getattr`; if the default is not given it raise an error, while if a default is given, that is returned instead.
+This design is to make it ergonomic both when all variables are expected to be present (one need not check manually that an actual `Var` was returned; it's sufficient to _not_ catch the exception), and when an action should be taken _only_ if the variable is present, such as in
+```python
+if (var := qc.get_var("a", None)) is not None:
+    # Do something with `var`, which is guaranteed to be a `Var`,
+    # but no extra `try`/`except` is needed if it might not exist.
+```
+These methods are present to solve a common user paper-cut when working with `Parameter` values, where they can't retrieve the parameter only by its string name despite circuits enforcing that these instances have unique names or are referentially identical within a circuit.
+
+The iterator methods `iter_vars`, `iter_declared_vars`, `iter_input_vars` and `iter_captured_vars` all return single-use iterator objects over the relevant variables.
+This choice lets `QuantumCircuit` use a mutable type internally for performance when adding new variables, while allowing iteration through the variables without the necessity of copying the collection when outputting it to the user to avoid them messing with it.
+
+
+### Examples
+
+The example given above in the motivation section would look like:
+
+```python
+from qiskit.circuit import QuantumCircuit
+from qiskit.circuit.classical import expr, types
+
+qc = QuantumCircuit(2)
+twirl_bits_0 = qc.add_input("twirl_bits_0", types.Bool())
+twirl_bits_1 = qc.add_input("twirl_bits_1", types.Bool())
+with qc.if_test(twirl_bits_0):
+    qc.x(0)
+with qc.if_test(twirl_bits_1):
+    qc.x(1)
+```
+
+The result of `qasm3.dumps()` on this circuit will be (subject to header-file discussions, etc):
+```qasm3
+OPENQASM 3.0;
+
+include 'stdgates.inc';
+
+input bool twirl_bits_0;
+input bool twirl_bits_1;
+
+qubit _qubit_0;
+qubit _qubit_1;
+
+if (twirl_bits_0) {
+    x _qubit_0;
+}
+if (twirl_bits_1) {
+    x _qubit_1;
+}
+```
+
+---
+
+Using the control-flow builders will allow scoped declarations, and using a variable declared in an outer builder scope will automatically cause it to become captured within the output.
+For example:
+
+```python
+from qiskit.circuit import QuantumCircuit, QuantumRegister, ClassicalRegister
+from qiskit.circuit.classical import expr, types
+
+qr = QuantumRegister(3)
+cr = ClassicalRegister(3)
+qc = QuantumCircuit(qr, cr)
+mask = qc.add_var("mask", types.Uint(3), expr.value(5))
```

Comment:
I find this a little hard to scan as an assignment `uint[3] mask = 5`, but I am struggling to come up with a good alternative. I guess it does read a little better when providing the `initial=` keyword.

Reply:
Yeah, it's not perfect for sure. This RFC is principally meant to be the low-level methods and implementation - I think a higher order interface would be a follow up.